Server: Broken order of Netty writes

### Original title: Not all JSONObjects get always sent

## Environment Details
* Helidon Version: 2.4.0
* Helidon SE
* JDK version:
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06, mixed mode, sharing)
* OS:
Distributor ID:	Ubuntu
Description:	Ubuntu 21.10
Release:	21.10
Codename:	impish
----------

## Problem Description
JsonObjects supplied with `Flow.Publisher<JsonObject>` are not always sent all. Happens intermittently.  

```java
response.send(Multi.range(0, 2)
                .flatMapCompletionStage(i -> CompletableFuture.supplyAsync(() -> JSON.createObjectBuilder().add("test", "kec").build()))
                .log(), JsonObject.class);
```

```shell
kec@romulus:~$ curl http://localhost:8080/updates
[{"test":"kec"}]
```

```
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onSubscribe(...)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇗ request(32)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onNext({"test":"kec"})
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇗ request(1)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onNext({"test":"kec"})
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇗ request(1)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onComplete()
```

Extra `0` byte introduced by `Charset.encode` in `io.helidon.media.jsonp.JsonpBodyWriter.JsonStructureToChunks#map` breaks Netty:
```
{"test":"kec0"} 
[123, 34, 116, 101, 115, 116, 34, 58, 34, 107, 101, 99, 48, 34, 125, 0]
java.nio.HeapByteBuffer[pos=0 lim=15 cap=16]
```
It should not be a problem because `ByteBuffer.limit=15` eg extra `0` should not be considered

But netty channel future ends exceptionally(error is swallowed):
```
io.netty.handler.codec.EncoderException: java.lang.IllegalStateException: unexpected message type: DefaultHttpContent, state: 0
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:104)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764)
	at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.IllegalStateException: unexpected message type: DefaultHttpContent, state: 0
	at io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:124)
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:89)
	... 11 more 
```

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -66,7 +66,7 @@ class BareResponseImpl implements BareResponse {
     private static final SocketClosedException CLOSED = new SocketClosedException("Response channel is closed!");
 
     private final boolean keepAlive;
-    private final ChannelHandlerContext ctx;
+    private final NettyChannel channel;
     private final AtomicBoolean statusHeadersSent = new AtomicBoolean(false);
     private final AtomicBoolean internallyClosed = new AtomicBoolean(false);
     private final CompletableFuture<BareResponse> responseFuture;
@@ -107,7 +107,7 @@ class BareResponseImpl implements BareResponse {
         this.requestEntityAnalyzed = requestEntityAnalyzed;
         this.responseFuture = new CompletableFuture<>();
         this.headersFuture = new CompletableFuture<>();
-        this.ctx = ctx;
+        this.channel = new NettyChannel(ctx);
         this.requestId = requestId;
         this.keepAlive = HttpUtil.isKeepAlive(request);
         this.requestHeaders = request.headers();
@@ -195,7 +195,7 @@ class BareResponseImpl implements BareResponse {
         // If already set (e.g. WebSocket upgrade), do not override
         if (keepAlive) {
             if (!requestContext.requestCompleted()) {
-                LOGGER.finer(() -> log("Request content not fully read with keep-alive: true", ctx));
+                LOGGER.finer(() -> log("Request content not fully read with keep-alive: true", channel));
                 if (!requestContext.hasRequests() || requestContext.requestCancelled()) {
                     requestEntityAnalyzed = requestEntityAnalyzed.thenApply(listener -> {
                         if (listener.equals(ChannelFutureListener.CLOSE)) {
@@ -207,7 +207,7 @@ class BareResponseImpl implements BareResponse {
                     });
                     //We are not sure which Connection header value should be set.
                     //If unhandled entity is only one content large, we can keep the keep-alive
-                    ctx.channel().read();
+                    channel.read();
                 } else {
                     response.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
                     requestEntityAnalyzed.complete(ChannelFutureListener.CLOSE);
@@ -298,11 +298,11 @@ class BareResponseImpl implements BareResponse {
             requestContext.runInScope(() -> {
                 if (ChannelFutureListener.CLOSE.equals(listener)) {
                     if (LOGGER.isLoggable(Level.FINEST)) {
-                        LOGGER.finest(log("Closing with an empty buffer; keep-alive: false", ctx));
+                        LOGGER.finest(log("Closing with an empty buffer; keep-alive: false", channel));
                     }
                 } else {
                     LOGGER.finest(() -> log("Writing an empty last http content; keep-alive: true"));
-                    ctx.channel().read();
+                    channel.read();
                 }
                 writeLastContent(throwable, listener);
             });
@@ -348,7 +348,7 @@ class BareResponseImpl implements BareResponse {
             }
         }
 
-        writeOnEventLoop(ctx, true, lastHttpContent)
+        channel.write(true, lastHttpContent)
                 .thenApply(f -> f
                         .addListener(completeOnFailureListener("An exception occurred when writing last http content."))
                         .addListener(completeOnSuccessListener(throwable))
@@ -368,7 +368,7 @@ class BareResponseImpl implements BareResponse {
         return future -> {
             if (future.isSuccess()) {
                 completeResponseFuture(throwable);
-                LOGGER.finest(() -> log("Last http message flushed", ctx));
+                LOGGER.finest(() -> log("Last http message flushed", channel));
             }
         };
     }
@@ -390,9 +390,9 @@ class BareResponseImpl implements BareResponse {
             requestContext.runInScope(() -> {
                 if (data.isFlushChunk()) {
                     if (prevRequestChunk == null) {
-                        flushOnEventLoop(ctx.flush());
+                        channel.flush();
                     } else {
-                        prevRequestChunk = prevRequestChunk.thenRun(() -> flushOnEventLoop(ctx));
+                        prevRequestChunk = prevRequestChunk.thenRun(channel::flush);
                     }
                     subscription.request(1);
                     return;
@@ -429,34 +429,33 @@ class BareResponseImpl implements BareResponse {
      * inside an {@link #orderedWrite(Runnable)} runnable.
      */
     private CompletionStage<ChannelFuture> initWriteResponse() {
-        CompletionStage<ChannelFuture> eventLoopDone = writeOnEventLoop(ctx, false, response)
+        CompletionStage<ChannelFuture> eventLoopDone = channel.write(false, response)
                 .thenApply(f -> f
-                        .addListener(future -> {
-                            if (future.isSuccess()) {
-                                headersFuture.complete(this);
-                            }
-                        })
+                        .addListener(future -> NettyChannel.completeFuture(future, headersFuture, this))
                         .addListener(completeOnFailureListener("An exception occurred when writing headers."))
                         .addListener(ChannelFutureListener.CLOSE_ON_FAILURE));
+
         response = null;
-
-
-            eventLoopDone = eventLoopDone.whenComplete((f, e) -> {
-                if (firstChunk != null) {
-                    sendData(firstChunk, false);
-                    firstChunk = null;
-                }
-            });
-
         lengthOptimization = false;
-        return eventLoopDone;
+
+        return eventLoopDone.whenComplete((f, e) -> {
+            if (firstChunk != null) {
+                sendData(firstChunk, false);
+                firstChunk = null;
+            }
+        });
     }
 
     /**
      * Submits a data chunk for writing. This method must be called inside an
      * {@link #orderedWrite(Runnable)} runnable.
      *
+     * When DataChunk has flush true, upstream is requested for another chunk
+     * after actual write is done. When flush is false, upstream is requested
+     * after request for write is finished.
+     *
      * @param data the chunk.
+     * @param requestOneMore if write should request for another chunk
      */
     private CompletionStage<ChannelFuture> sendData(DataChunk data, boolean requestOneMore) {
         LOGGER.finest(() -> log("Sending data chunk"));
@@ -477,25 +476,25 @@ class BareResponseImpl implements BareResponse {
             httpContent = new DefaultHttpContent(Unpooled.wrappedBuffer(data.data()));
         }
 
-        LOGGER.finest(() -> log("Sending data chunk on event loop thread", ctx));
+        LOGGER.finest(() -> log("Sending data chunk on event loop thread", channel));
 
-        return writeOnEventLoop(ctx, data.flush(), httpContent)
+        return channel.write(data.flush(), httpContent)
                 .thenApply(f -> {
+                    // After request for write is made on event loop thread
                     if (!data.flush() && requestOneMore) {
+                        // No flush, request another chunk immediately, this one needs to wait in the cache
                         subscription.request(1);
                     }
+                    // Add listeners to execute when actual write is done
                     return f.addListener(future -> {
-                                data.writeFuture().ifPresent(writeFuture -> {
-                                    // Complete write future based con channel future
-                                    if (future.isSuccess()) {
-                                        writeFuture.complete(data);
-                                    } else {
-                                        writeFuture.completeExceptionally(future.cause());
-                                    }
-                                });
+                                // Complete write future based con channel future
+                                data.writeFuture()
+                                        .ifPresent(writeFuture -> NettyChannel.completeFuture(future, writeFuture, data));
+
                                 boolean flush = data.flush();
                                 data.release();
                                 if (flush && requestOneMore) {
+                                    // Previous chunk sent, request another one
                                     subscription.request(1);
                                 }
                                 LOGGER.finest(() -> log("Data chunk sent with result: %s", future.isSuccess()));
@@ -504,36 +503,6 @@ class BareResponseImpl implements BareResponse {
                             .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
                 });
     }
-
-    private void flushOnEventLoop(ChannelHandlerContext ctx){
-        if (ctx.executor().inEventLoop()){
-            ctx.flush();
-        } else {
-            ctx.executor().execute(ctx::flush);
-        }
-    }
-
-    private CompletionStage<ChannelFuture> writeOnEventLoop(ChannelHandlerContext ctx, boolean flush, Object msg) {
-        CompletableFuture<ChannelFuture> channelFuture = new CompletableFuture<>();
-        if (ctx.executor().inEventLoop()) {
-            if (flush) {
-                channelFuture.complete(ctx.writeAndFlush(msg));
-            } else {
-                channelFuture.complete(ctx.write(msg));
-            }
-        } else {
-            ctx.executor().execute(() -> {
-                if (flush) {
-                    channelFuture.complete(ctx.writeAndFlush(msg));
-                } else {
-                    channelFuture.complete(ctx.write(msg));
-                }
-            });
-        }
-
-        return channelFuture;
-    }
-
 
     @Override
     public void onError(Throwable thr) {
@@ -590,7 +559,7 @@ class BareResponseImpl implements BareResponse {
     private String log(String template, Object... params) {
         List<Object> list = new ArrayList<>(params.length + 3);
         list.add(System.identityHashCode(this));
-        list.add(ctx != null ? ctx.channel().id() : "N/A");
+        list.add(channel.id());
         list.add(http2StreamId != null ? http2StreamId : "N/A");
         list.addAll(Arrays.asList(params));
         return String.format("[Response: %s, Channel: 0x%s, StreamID: %s] " + template, list.toArray());

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -80,7 +80,7 @@ class BareResponseImpl implements BareResponse {
 
     // Accessed by Subscriber method threads
     private Flow.Subscription subscription;
-    private DataChunk firstChunk;
+    private volatile DataChunk firstChunk;
     private CompletableFuture<?> prevRequestChunk;
     private CompletableFuture<ChannelFutureListener> requestEntityAnalyzed;
 
@@ -416,19 +416,20 @@ class BareResponseImpl implements BareResponse {
      *
      * @param data the data chunk.
      */
-    private void onNextPipe(DataChunk data) {
+    private CompletionStage<ChannelFuture> onNextPipe(DataChunk data) {
         if (lengthOptimization) {
-            initWriteResponse();
+            return initWriteResponse().thenCompose(f -> sendData(data, true));
+        } else {
+            return sendData(data, true);
         }
-        sendData(data);
     }
 
     /**
      * Initiates write of response and sends first chunk if available. This method must be called
      * inside an {@link #orderedWrite(Runnable)} runnable.
      */
-    private void initWriteResponse() {
-        writeOnEventLoop(ctx, false, response)
+    private CompletionStage<ChannelFuture> initWriteResponse() {
+        CompletionStage<ChannelFuture> eventLoopDone = writeOnEventLoop(ctx, false, response)
                 .thenApply(f -> f
                         .addListener(future -> {
                             if (future.isSuccess()) {
@@ -438,11 +439,17 @@ class BareResponseImpl implements BareResponse {
                         .addListener(completeOnFailureListener("An exception occurred when writing headers."))
                         .addListener(ChannelFutureListener.CLOSE_ON_FAILURE));
         response = null;
-        if (firstChunk != null) {
-            sendData(firstChunk);
-            firstChunk = null;
-        }
+
+
+            eventLoopDone = eventLoopDone.whenComplete((f,e) -> {
+                if (firstChunk != null) {
+                    sendData(firstChunk, false);
+                    firstChunk = null;
+                }
+            });
+
         lengthOptimization = false;
+        return eventLoopDone;
     }
 
     /**
@@ -451,7 +458,7 @@ class BareResponseImpl implements BareResponse {
      *
      * @param data the chunk.
      */
-    private void sendData(DataChunk data) {
+    private CompletionStage<ChannelFuture> sendData(DataChunk data, boolean requestOneMore) {
         LOGGER.finest(() -> log("Sending data chunk"));
 
         DefaultHttpContent httpContent;
@@ -472,9 +479,9 @@ class BareResponseImpl implements BareResponse {
 
         LOGGER.finest(() -> log("Sending data chunk on event loop thread", ctx));
 
-        Single.create(writeOnEventLoop(ctx, data.flush(), httpContent)
+        return writeOnEventLoop(ctx, data.flush(), httpContent)
                 .thenApply(f -> {
-                    if (!data.flush()) {
+                    if (!data.flush() && requestOneMore) {
                         subscription.request(1);
                     }
                     return f.addListener(future -> {
@@ -488,14 +495,14 @@ class BareResponseImpl implements BareResponse {
                                 });
                                 boolean flush = data.flush();
                                 data.release();
-                                if (flush) {
+                                if (flush && requestOneMore) {
                                     subscription.request(1);
                                 }
                                 LOGGER.finest(() -> log("Data chunk sent with result: %s", future.isSuccess()));
                             })
                             .addListener(completeOnFailureListener("Failure when sending a content!"))
                             .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
-                })).await(); // FIXME: This is a problem!
+                });
     }
 
     private void flushOnEventLoop(ChannelHandlerContext ctx){

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -22,6 +22,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
 import java.util.concurrent.Flow;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.logging.Level;
@@ -346,10 +347,12 @@ class BareResponseImpl implements BareResponse {
                 LOGGER.severe(() -> log("Upstream error while sending response: %s", throwable));
             }
         }
-        ctx.writeAndFlush(lastHttpContent)
-                .addListener(completeOnFailureListener("An exception occurred when writing last http content."))
-                .addListener(completeOnSuccessListener(throwable))
-                .addListener(closeAction);
+
+        writeOnEventLoop(ctx, true, lastHttpContent)
+                .thenApply(f -> f
+                        .addListener(completeOnFailureListener("An exception occurred when writing last http content."))
+                        .addListener(completeOnSuccessListener(throwable))
+                        .addListener(closeAction));
     }
 
     private GenericFutureListener<Future<? super Void>> completeOnFailureListener(String message) {
@@ -387,9 +390,9 @@ class BareResponseImpl implements BareResponse {
             requestContext.runInScope(() -> {
                 if (data.isFlushChunk()) {
                     if (prevRequestChunk == null) {
-                        ctx.flush();
+                        flushOnEventLoop(ctx.flush());
                     } else {
-                        prevRequestChunk = prevRequestChunk.thenRun(ctx::flush);
+                        prevRequestChunk = prevRequestChunk.thenRun(() -> flushOnEventLoop(ctx));
                     }
                     subscription.request(1);
                     return;
@@ -423,25 +426,23 @@ class BareResponseImpl implements BareResponse {
     /**
      * Initiates write of response and sends first chunk if available. This method must be called
      * inside an {@link #orderedWrite(Runnable)} runnable.
-     *
-     * @return Future of response or first chunk.
      */
-    private ChannelFuture initWriteResponse() {
-        ChannelFuture cf = ctx.write(response)
-                .addListener(future -> {
-                    if (future.isSuccess()) {
-                        headersFuture.complete(this);
-                    }
-                })
-                .addListener(completeOnFailureListener("An exception occurred when writing headers."))
-                .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+    private void initWriteResponse() {
+        writeOnEventLoop(ctx, false, response)
+                .thenApply(f -> f
+                        .addListener(future -> {
+                            if (future.isSuccess()) {
+                                headersFuture.complete(this);
+                            }
+                        })
+                        .addListener(completeOnFailureListener("An exception occurred when writing headers."))
+                        .addListener(ChannelFutureListener.CLOSE_ON_FAILURE));
         response = null;
         if (firstChunk != null) {
-            cf = sendData(firstChunk);
+            sendData(firstChunk);
             firstChunk = null;
         }
         lengthOptimization = false;
-        return cf;
     }
 
     /**
@@ -449,9 +450,8 @@ class BareResponseImpl implements BareResponse {
      * {@link #orderedWrite(Runnable)} runnable.
      *
      * @param data the chunk.
-     * @return channel future.
      */
-    private ChannelFuture sendData(DataChunk data) {
+    private void sendData(DataChunk data) {
         LOGGER.finest(() -> log("Sending data chunk"));
 
         DefaultHttpContent httpContent;
@@ -472,33 +472,59 @@ class BareResponseImpl implements BareResponse {
 
         LOGGER.finest(() -> log("Sending data chunk on event loop thread", ctx));
 
-        ChannelFuture channelFuture;
-        if (data.flush()) {
-            channelFuture = ctx.writeAndFlush(httpContent);
-        } else {
-            channelFuture = ctx.write(httpContent);
-            subscription.request(1);
-        }
-
-        return channelFuture
-                .addListener(future -> {
-                    data.writeFuture().ifPresent(writeFuture -> {
-                        // Complete write future based con channel future
-                        if (future.isSuccess()) {
-                            writeFuture.complete(data);
-                        } else {
-                            writeFuture.completeExceptionally(future.cause());
-                        }
-                    });
-                    boolean flush = data.flush();
-                    data.release();
-                    if (flush) {
+        Single.create(writeOnEventLoop(ctx, data.flush(), httpContent)
+                .thenApply(f -> {
+                    if (!data.flush()) {
                         subscription.request(1);
                     }
-                    LOGGER.finest(() -> log("Data chunk sent with result: %s", future.isSuccess()));
-                })
-                .addListener(completeOnFailureListener("Failure when sending a content!"))
-                .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+                    return f.addListener(future -> {
+                                data.writeFuture().ifPresent(writeFuture -> {
+                                    // Complete write future based con channel future
+                                    if (future.isSuccess()) {
+                                        writeFuture.complete(data);
+                                    } else {
+                                        writeFuture.completeExceptionally(future.cause());
+                                    }
+                                });
+                                boolean flush = data.flush();
+                                data.release();
+                                if (flush) {
+                                    subscription.request(1);
+                                }
+                                LOGGER.finest(() -> log("Data chunk sent with result: %s", future.isSuccess()));
+                            })
+                            .addListener(completeOnFailureListener("Failure when sending a content!"))
+                            .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+                })).await(); // FIXME: This is a problem!
+    }
+
+    private void flushOnEventLoop(ChannelHandlerContext ctx){
+        if (ctx.executor().inEventLoop()){
+            ctx.flush();
+        } else {
+            ctx.executor().execute(ctx::flush);
+        }
+    }
+
+    private CompletionStage<ChannelFuture> writeOnEventLoop(ChannelHandlerContext ctx, boolean flush, Object msg) {
+        CompletableFuture<ChannelFuture> channelFuture = new CompletableFuture<>();
+        if (ctx.executor().inEventLoop()) {
+            if (flush) {
+                channelFuture.complete(ctx.writeAndFlush(msg));
+            } else {
+                channelFuture.complete(ctx.write(msg));
+            }
+        } else {
+            ctx.executor().execute(() -> {
+                if (flush) {
+                    channelFuture.complete(ctx.writeAndFlush(msg));
+                } else {
+                    channelFuture.complete(ctx.write(msg));
+                }
+            });
+        }
+
+        return channelFuture;
     }
 
 

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -441,7 +441,7 @@ class BareResponseImpl implements BareResponse {
         response = null;
 
 
-            eventLoopDone = eventLoopDone.whenComplete((f,e) -> {
+            eventLoopDone = eventLoopDone.whenComplete((f, e) -> {
                 if (firstChunk != null) {
                     sendData(firstChunk, false);
                     firstChunk = null;

--- a/webserver/webserver/src/main/java/io/helidon/webserver/NettyChannel.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/NettyChannel.java
@@ -1,0 +1,131 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.webserver;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelId;
+import io.netty.util.concurrent.Future;
+
+/**
+ * Wrapper for {@link io.netty.channel.ChannelHandlerContext} guarding all writes to be made exclusively from event loop thread.
+ *
+ * <ul>
+ * <b>Netty can provide "ordering" in the one of the following situations exclusively:</b>
+ *  <ul>
+ *      <li>You are doing all writes from the EventLoop thread.</li>
+ *      <li>You are doing no writes from the EventLoop thread (i.e. all writes are being done in other thread(s)).</li>
+ *  </ul>
+ * </ul>
+ *
+ * @see <a href="https://github.com/netty/netty/issues/3887#issuecomment-112540327"
+ * >https://github.com/netty/netty/issues/3887#issuecomment-112540327</a>
+ */
+class NettyChannel {
+    private final ChannelHandlerContext ctx;
+
+    NettyChannel(ChannelHandlerContext ctx) {
+        this.ctx = ctx;
+    }
+
+    /**
+     * Returns the globally unique identifier of the underlying Channel.
+     *
+     * @return globally unique identifier
+     */
+    ChannelId id() {
+        return ctx.channel().id();
+    }
+
+    /**
+     * Request to Read data from the Channel into the first inbound buffer,
+     * triggers an ChannelInboundHandler.channelRead(ChannelHandlerContext, Object) event if data was read, and triggers
+     * a channelReadComplete event so the handler can decide to continue reading.
+     * If there's a pending read operation already, this method does nothing.
+     * This will result in having the ChannelOutboundHandler.read(ChannelHandlerContext)
+     * method called of the next ChannelOutboundHandler contained in the ChannelPipeline of the Channel.
+     */
+    void read() {
+        ctx.read();
+    }
+
+    /**
+     * Request to flush all pending messages via this ChannelOutboundInvoker from Netty's event loop thread.
+     */
+    void flush() {
+        if (ctx.executor().inEventLoop()) {
+            ctx.flush();
+        } else {
+            ctx.executor().execute(ctx::flush);
+        }
+    }
+
+    /**
+     * Request to write a message via this ChannelHandlerContext through the ChannelPipeline from Netty's event loop thread.
+     *
+     * @param flush flush immediately
+     * @param msg   message to write
+     * @return CompletionStage completed when request is made from event loop thread, containing future of actual write
+     */
+    CompletionStage<ChannelFuture> write(boolean flush, Object msg) {
+        CompletableFuture<ChannelFuture> channelFuture = new CompletableFuture<>();
+
+        if (ctx.executor().inEventLoop()) {
+            // Fast path for items emitted by event loop thread
+            if (flush) {
+                channelFuture.complete(ctx.writeAndFlush(msg));
+            } else {
+                channelFuture.complete(ctx.write(msg));
+            }
+        } else {
+            if (flush) {
+                ctx.executor().execute(() -> channelFuture.complete(ctx.writeAndFlush(msg)));
+            } else {
+                ctx.executor().execute(() -> channelFuture.complete(ctx.write(msg)));
+            }
+        }
+
+        return channelFuture;
+    }
+
+    /**
+     * Map Netty's future completing with void to CompletableFuture completing with supplied item.
+     *
+     * @param nettyFuture Netty's future completing with void
+     * @param completable CompletableFuture completing with supplied item
+     * @param item        to complete supplied CompletableFuture with
+     * @param <T>         type of the CompletableFuture's item
+     */
+    static <T> void completeFuture(Future<? super Void> nettyFuture, CompletableFuture<T> completable, T item) {
+        if (nettyFuture.isSuccess()) {
+            completable.complete(item);
+        } else {
+            completable.completeExceptionally(nettyFuture.cause());
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "NettyChannel{"
+                + "ctx=" + ctx.toString()
+                + '}';
+    }
+}

--- a/webserver/webserver/src/test/java/io/helidon/webserver/OrderOfWritesTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/OrderOfWritesTest.java
@@ -5,7 +5,7 @@
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- * http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,

--- a/webserver/webserver/src/test/java/io/helidon/webserver/OrderOfWritesTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/OrderOfWritesTest.java
@@ -17,10 +17,6 @@
 
 package io.helidon.webserver;
 
-import java.net.URI;
-import java.net.http.HttpClient;
-import java.net.http.HttpRequest;
-import java.net.http.HttpResponse;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.time.Duration;
@@ -30,7 +26,10 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import io.helidon.common.http.DataChunk;
+import io.helidon.common.http.Http;
 import io.helidon.common.reactive.Multi;
+import io.helidon.webclient.WebClient;
+import io.helidon.webclient.WebClientResponse;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
@@ -71,22 +70,25 @@ public class OrderOfWritesTest {
                     .start()
                     .await(TIME_OUT);
 
-            HttpClient client = HttpClient.newBuilder()
-                    .version(HttpClient.Version.HTTP_1_1)
-                    .followRedirects(HttpClient.Redirect.NORMAL)
-                    .connectTimeout(TIME_OUT)
-                    .build();
-
-            HttpRequest request = HttpRequest.newBuilder()
-                    .uri(URI.create("http://localhost:"+server.port()))
-                    .timeout(TIME_OUT)
-                    .GET()
+            WebClient client = WebClient.builder()
+                    .baseUri("http://localhost:" + server.port())
+                    .connectTimeout(TIME_OUT.toMillis(), TimeUnit.MILLISECONDS)
                     .build();
 
             for (AtomicInteger i = new AtomicInteger(); i.get() < 5000; i.getAndIncrement()) {
-                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
-                String s = response.body();
-                Assertions.assertEquals(expected, s, "Failed at " + i.get() + " " + s);
+                WebClientResponse response = null;
+                try {
+                    response = client.get()
+                            .request()
+                            .await(TIME_OUT);
+                    Assertions.assertEquals(Http.ResponseStatus.Family.SUCCESSFUL, response.status().family());
+                    String content = response.content().as(String.class).await(TIME_OUT);
+                    Assertions.assertEquals(expected, content, "Failed at " + i.get() + " " + content);
+                } finally {
+                    if (response != null) {
+                        response.close();
+                    }
+                }
             }
         } finally {
             exec.shutdownNow();

--- a/webserver/webserver/src/test/java/io/helidon/webserver/OrderOfWritesTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/OrderOfWritesTest.java
@@ -1,0 +1,99 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.webserver;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.helidon.common.http.DataChunk;
+import io.helidon.common.reactive.Multi;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+public class OrderOfWritesTest {
+
+    private static final Duration TIME_OUT = Duration.ofSeconds(5);
+
+    @Test
+    void threadMixUp() throws Exception {
+        final String expected = "_1_2_3";
+        final byte[] underscore = "_".getBytes(StandardCharsets.UTF_8);
+        final ExecutorService exec = Executors.newSingleThreadExecutor();
+
+        WebServer server = null;
+
+        try {
+            server = WebServer.builder()
+                    .routing(Routing.builder()
+                            .get((req, res) -> res.send(Multi.just("1", "2", "3")
+                                    .map(String::valueOf)
+                                    .map(String::getBytes)
+                                    // Condition: initiate write from different thread than event loop
+                                    .observeOn(exec)
+                                    // Condition: mix-up writes with flushes
+                                    .flatMap(number -> Multi.just(
+                                            DataChunk.create(false, ByteBuffer.wrap(underscore)),
+                                            DataChunk.create(true, ByteBuffer.wrap(number))
+                                    ))
+                            ))
+                            .build())
+                    .host("localhost")
+                    .port(0)
+                    .build()
+                    .start()
+                    .await(TIME_OUT);
+
+            HttpClient client = HttpClient.newBuilder()
+                    .version(HttpClient.Version.HTTP_1_1)
+                    .followRedirects(HttpClient.Redirect.NORMAL)
+                    .connectTimeout(TIME_OUT)
+                    .build();
+
+            HttpRequest request = HttpRequest.newBuilder()
+                    .uri(URI.create("http://localhost:"+server.port()))
+                    .timeout(TIME_OUT)
+                    .GET()
+                    .build();
+
+            for (AtomicInteger i = new AtomicInteger(); i.get() < 5000; i.getAndIncrement()) {
+                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+                String s = response.body();
+                Assertions.assertEquals(expected, s, "Failed at " + i.get() + " " + s);
+            }
+        } finally {
+            exec.shutdownNow();
+            if (server != null) {
+                server.shutdown().await(TIME_OUT);
+            }
+            assertThat(exec.awaitTermination(TIME_OUT.toMillis(), TimeUnit.MILLISECONDS), is(true));
+        }
+    }
+}

